classes/viewsessions/viewsessions_table.php

```diff
@@ -0,0 +1,163 @@
+            'observee_username',
+            'state',
+            'start_time',
+            'finish_time',
+            'action'
+        ]);
+
+        $this->define_headers([
+            get_string('id', 'observation'),
+            get_string('observer', 'observation'),
+            get_string('observee', 'observation'),
+            get_string('state', 'observation'),
+            get_string('starttime', 'observation'),
+            get_string('finishtime', 'observation'),
+            get_string('actions', 'observation'),
+        ]);
+
+        $this->pagesize = $perpage;
+
+        $systemcontext = \context_system::instance();
+        $this->context = $systemcontext;
+        $this->collapsible(false);
+        $this->pageable(true);
+        $this->is_downloadable(false);
+        $this->define_baseurl($callbackurl);
+        $this->no_sorting('action');
+    }
+
+    /**
+     * Data converter for the state column
+     * @param mixed $row current row
+     */
+    public function col_state($row) {
+        return get_string($row->state, 'observation');
+    }
+
+    /**
+     * Data converter for the finish_time column
+     * @param mixed $row current row
+     */
+    public function col_finish_time($row) {
+        return $this->format_as_date($row->finish_time);
+    }
+
+    /**
+     * Data converter for the start_time column
+     * @param mixed $row current row
+     */
+    public function col_start_time($row) {
+        return $this->format_as_date($row->start_time);
+    }
+
+    /**
+     * Formats UNIX epoch as date for date columns in table.
+     * @param mixed $epochs UNIX epoch time or null
+     */
+    private function format_as_date($epochs) {
+        if (!is_null($epochs)) {
+            return date('r', $epochs);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Creates an action button for the col_action function.
+     * @param \moodle_url $url URL for the button to link to
+     * @param string $text button text
+     * @param string $style button classes
+     */
+    private function action_button(\moodle_url $url, string $text, string $style = '') {
+        return \html_writer::link(
+            $url,
+            $text,
+            ['class' => 'btn mr-2 '.$style]
+        );
+    }
+
+    /**
+     * Data converter for the action column
+     * @param mixed $row current row
+     */
+    public function col_action($row) {
+        // Add action buttons.
+        $htmlout = '';
+
+        if ($row->state === 'complete') {
+            // View summary button.
+            $htmlout .= $this->action_button(
+                new \moodle_url('sessionsummary.php', ['sessionid' => $row->id, 'mode' => 'viewing']),
+                get_string('viewsummary', 'observation'),
+                'btn-info');
+
+            // Re-open button.
+            $htmlout .= $this->action_button(
+                new \moodle_url('session.php', ['sessionid' => $row->id]),
```

Comment:
just to be safe incase of php being naughty, I like to generally namespace this completely, eg

`new moodle_url('/mod/observation/session.php', ...`

PHP has historically had some weird behaviour with relative pathing, not sure how that interacts with moodle_url